### ISSUE TITLE
Trying to move axios to System.js import map, and having tons of problems

### ISSUE THREAD

**Issue by craigmiller160** [react-microfrontends/api#1]
Hi there. I'm not sure if this is the appropriate forum to raise this question, but I'm attempting to learn how single-spa works. One of my biggest areas of interest is dependency sharing between micro-frontends. This is because my company uses a homegrown micro-frontend framework, and dependency duplication is the biggest limitation of our implementation. So I want to really see how far this dependency sharing stuff here can go, by migrating all shared dependencies in your examples to come from SystemJS import maps.

Anyway, when I do this for axios, I end up with the following error (I un-minified everything to get better messages):

`Uncaught TypeError: application '@react-mf/people' died in status LOADING_SOURCE_CODE: axios__WEBPACK_IMPORTED_MODULE_2___default.a.create is not a function`

This error is coming from line 8 in fetchWithCache.js, where the axiosInstance is created. So, let's cover how I moved axios to SystemJS import map, and then move on from there. I'm attaching two files. One is the index.cjs from the root-config project. As you can see, I'm loading two source maps. One contains the library dependencies, the other contains the paths to my projects. I will attach both source maps, and the webpack config as well.

If you could point me in the right direction, I would appreciate it. Thank you.\

Edit: Ok, so I can't upload the files. So I will include the code samples here:

index.ejs from root-config
```
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <meta http-equiv="X-UA-Compatible" content="ie=edge">
  <title>Root Config</title>
  <meta name="importmap-type" content="systemjs-importmap" />
  <!-- Shared Dependencies Import Map -->
  <script type="systemjs-importmap" src="http://localhost:9002/importmap.json"></script>
  <script type="systemjs-importmap" src="http://localhost:9001/app-importmap.json"></script>
  <script src="https://cdn.jsdelivr.net/npm/import-map-overrides/dist/import-map-overrides.js"></script>
  <script src="https://cdn.jsdelivr.net/npm/systemjs/dist/system.min.js"></script>
  <script src="https://cdn.jsdelivr.net/npm/systemjs/dist/extras/amd.min.js"></script>
  <script src="https://cdn.jsdelivr.net/npm/systemjs/dist/extras/named-exports.min.js"></script>
  <script src="https://cdn.jsdelivr.net/npm/systemjs/extras/use-default.js"></script>
</head>
<body>
  <script>
    System.import('@react-mf/root-config');
    System.import('@react-mf/styleguide');
  </script>
  <import-map-overrides-full show-when-local-storage="devtools" dev-libs></import-map-overrides-full>
</body>
</html>
```

webpack.config from api

```
const webpackMerge = require("webpack-merge");
const singleSpaDefaults = require("webpack-config-single-spa-react");

module.exports = webpackConfigEnv => {
  const defaultConfig = singleSpaDefaults({
    orgName: "react-mf",
    projectName: "api",
    webpackConfigEnv
  });

  return webpackMerge.smart(defaultConfig, {
    externals: [
      "axios"
    ]
    // modify the webpack config however you'd like to by adding to this object
  });
};
```

dependencies importmap

```
{
  "imports": {
    "single-spa": "https://cdn.jsdelivr.net/npm/single-spa@5.0.0-beta.1/lib/system/single-spa.min.js",
    "react": "https://cdn.jsdelivr.net/npm/react@16.12.0/umd/react.production.min.js",
    "react-dom": "https://cdn.jsdelivr.net/npm/react-dom@16.12.0/umd/react-dom.production.min.js",
    "axios": "https://cdnjs.cloudflare.com/ajax/libs/axios/0.19.2/axios.min.js"
  }
}
```

**Comment by frehner** [comment]
First impression is that import maps need to be defined before system js is script tagged. Try that and let me know if anything changes

**Comment by craigmiller160** [comment]
> First impression is that import maps need to be defined before system js is script tagged. Try that and let me know if anything changes

Nope, didn't help. I updated my OP to reflect my new index.ejs. I should add that the axios JS file is definitely loading, I can see it in the network tab. It's just not being exposed properly.

**Comment by internettrans** [comment]
Here's a code sandbox that shows axios being loaded properly - https://codesandbox.io/s/rough-meadow-b6o70

One thing to note is that it is being loaded as a module that has a `default` export, but no named exports.

**Works**
```js
import axios from 'axios'

axios.get('https://dog.ceo/api/breeds/image/random');
```

**But the following doesn't**
```js
import { get } from 'axios'

get('https://dog.ceo/api/breeds/image/random');
```

If that's the issue, I can show you workarounds. If it's not the issue, the next step I'd take to debug is to put a console.log() **inside of your webpack application** to see what the Axios object is:

```js
import * as axios from 'axios'

console.log('axios', axios);
```

You need to verify that the object logged has a `default` property that is a function with other functions on it as properties. It should **not** be this:

```js
{default: {default: function() {}}}
```

If you're getting the "double default" problem, try adding the following to your HTML page:

```js
<script src="https://cdn.jsdelivr.net/npm/systemjs/extras/use-default.js"></script>
```

**Comment by craigmiller160** [comment]
So I don't get double-default, but I do get default-as-a-function. I'm attaching a screenshot below of what I saw when I console.logged axios. I did the `import * as axios from 'axios' first`.

![axios](https://user-images.githubusercontent.com/12954037/75775970-0f497e00-5d21-11ea-87c5-1e8579dd76f5.png)

Edit: Ok, maybe it is double-default? I poked at this for a few minutes, and I found a way to make the axios code work:

```
import axios from "axios";

const axiosInstance = axios.default.default.create({
  baseURL,
  timeout: 20000
});
```

So, clearly something strange is happening there. The question is, how do I fix it. I tried adding the use-default to my HTML page, I updated the example in the OP to include it.

Edit2: This code works as well,with a single default. Again, I'm using the use-default script, so maybe that's helping?

```
import axios from "axios";

const axiosInstance = axios.default.create({
  baseURL,
  timeout: 20000
});
```

**Comment by internettrans** [comment]
If you don't need the SystemJS AMD extra, remove it and that would get rid of the need to do the extra `default` property.

The reason for the default property is that SystemJS tries to interop with AMD modules but there are some nuances to that resulted in Guy Bedford and I putting the exported object under the `default` property.

If you remove the AMD extra, you'll fall back to "global loading" which will still work for axios and won't have the weird default problem.

**Comment by craigmiller160** [comment]
Interesting. So in that case you would say that it's hard to set up a configuration that successfully handles all possible dependencies?

Not trying to be critical here. Some quick background: my company's application uses a micro-frontend architecture, and we've been in production with it for almost a year. I'm the one who designed and built the architecture, it's based on some articles I read on how to use web components as the "glue", if you will, between the frontends. It's an effective approach that has worked well for us, but it has some limitations, the biggest one being that each micro-frontend needs a separate copy of all its dependencies, which has resulted in our total application size growing quite large.

Since I first read about your framework (or whatever you want to call this), I've been extremely fascinated by it. I am very impressed by what single-spa can do as-is, and I'm evaluating it in the context of a replacement for what I had built last year.

So is there some way to create a setup that can handle all possible dependencies?

**Comment by internettrans** [comment]
> So in that case you would say that it's hard to set up a configuration that successfully handles all possible dependencies?

It can be, yes, although it has gotten a lot better. I've submitted and published about 10 fixes to SystemJS in the last year that make it so that most people never have to worry about module format interop anymore.

> So is there some way to create a setup that can handle all possible dependencies?

Yes, but we need some PRs merged first. I just created https://github.com/webpack/webpack/pull/10503 (a PR I've been meaning to get around to for a while). I believe that that will solve the problem entirely for you. Please show your support there with thumbs up / comment.

You can put into place workarounds while waiting for it - I've been working for about a year to get seamless webpack / systemjs / amd interop and it's pretty close. The default export for webpack externals in bundles compiled to System.register format is one of the last things I'm aware of. I think that AMD wouldn't be a problem if that PR gets merged.

Here are a couple options that you can do for now as a workaround:

1. Create a repo under https://github.com/esm-bundle that has a systemjs version of axios. See [this blog post](https://medium.com/@joeldenning/an-esm-bundle-for-any-npm-package-5f850db0e04d) for more details
2. Add the following code to your index.html as a workaround (see [codepen that shows it works](https://codesandbox.io/s/gallant-chaplygin-fsjnf))
```html
<script src="system.js"></script>
<!-- all your other answers -->
<script>
  (function() {
    const originalInstantiate = System.constructor.prototype.instantiate;
    System.constructor.prototype.instantiate = function() {
      return originalInstantiate
        .apply(this, arguments)
        .then(function(register) {
          const originalFn = register[1];
          register[1] = function(_export, _context) {
            _export("__esModule", true);
            return originalFn.call(this, _export, _context);
          };
          return register;
        });
    };
  })();
</script>
```

3. Hacky workaround inside of your webpack bundle:
```js
// src/axios.js in your code
import axios from 'axios'

export default axios.default;
```

```js
// Whenever using axios, import it this way
import axios from 'src/axios';
```

**Comment by internettrans** [comment]
https://github.com/webpack/webpack/pull/10503 has been merged, and the workaround I gave you works as well. I am closing this. The webpack PR will be published in the next webpack@5 beta

**Comment by internettrans** [comment]
FYI this is now fixed in webpack@5.0.0-beta.15. See https://github.com/webpack/webpack/releases/tag/v5.0.0-beta.15 and https://github.com/webpack/webpack/pull/10503

**Comment by internettrans** [comment]
There was a bug in the custom systemjs extra above (see https://github.com/single-spa/single-spa-layout/issues/128). I have updated the comment with the corrected version
